app/src/main/java/org/mozilla/tv/firefox/fxa/FxaRepo.kt

```diff
@@ -0,0 +1,129 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.tv.firefox.fxa
+
+import android.content.Context
+import android.util.Log
+import androidx.lifecycle.ProcessLifecycleOwner
+import io.reactivex.subjects.BehaviorSubject
+import kotlinx.coroutines.Deferred
+import mozilla.appservices.fxaclient.Config
+import mozilla.components.concept.sync.AccountObserver
+import mozilla.components.concept.sync.DeviceCapability
+import mozilla.components.concept.sync.DeviceEvent
+import mozilla.components.concept.sync.DeviceEventsObserver
+import mozilla.components.concept.sync.DeviceType
+import mozilla.components.concept.sync.OAuthAccount
+import mozilla.components.concept.sync.Profile
+import mozilla.components.service.fxa.manager.DeviceTuple
+import mozilla.components.service.fxa.manager.FxaAccountManager
+import org.mozilla.tv.firefox.fxa.FxaRepo.AccountState.AUTHENTICATED
+import org.mozilla.tv.firefox.fxa.FxaRepo.AccountState.AUTHENTICATED_NO_PROFILE
+import org.mozilla.tv.firefox.fxa.FxaRepo.AccountState.NEEDS_REAUTHENTICATION
+import org.mozilla.tv.firefox.fxa.FxaRepo.AccountState.NOT_AUTHENTICATED
+
+private const val LOGTAG = "FxaRepo"
+
+private val APPLICATION_SCOPES = arrayOf(
+    // We don't use sync, however, if we later add the sync scope, already authenticated users
+    // will have to log in again. To avoid this in the future, we preemptively declare a sync scope.
+    "https://identity.mozilla.com/apps/oldsync"
+)
+
+/**
+ * Manages Firefox Account (FxA) state. In particular, [accountStateDontUseMeYet] exposes the current sign in
+ * state of the user. If you want to initiate sign in, see [FxaLoginUseCase.beginLogin].
+ *
+ * Devs should use this class rather than interacting with the FxA library directly.
+ */
+class FxaRepo(
+    context: Context,
+    val accountManager: FxaAccountManager = newInstanceDefaultAccountManager(context)
+) {
+
+    enum class AccountState {
+        // TODO: Are these states accurate?
+        // TODO: Later, may need "failed to login": https://github.com/mozilla-mobile/android-components/issues/3712
+        AUTHENTICATED,
+        AUTHENTICATED_NO_PROFILE, // todo: necessary? profile seems to be fetched async so yes?
+        NEEDS_REAUTHENTICATION,
+        NOT_AUTHENTICATED
+        // TODO: do we need initial state before state is fetched from disk?
+    }
+
+    // TODO: set state accurately, choose correct subject, set initial state, etc.
+    val accountStateDontUseMeYet = BehaviorSubject.create<AccountState>()
```

Comment:
FWIW, in both Fenix and the reference-browser an initialized `accountManager` is exposed to the rest of the application, and UI bits that care about account state can query it and/or register to receive updates. This works pretty well.

It seems to me like you'll end-up with the same pattern here, but via your own intermediary. I'm not sure what value this adds (versus exposing `accountManager`), but perhaps it makes things easier for you.

My general recommendation would be to avoid modelling this after internals of the account manager, but rather build on top of data abstractions that it provides (e.g. we're logged in; we have some auth problems; we have a new version of the profile; etc). Generally speaking, any module you add on top of some other module should ideally introduce a higher level of abstraction. If it doesn't, there's a good chance that module isn't needed and just adds complexity.

Reply:
@grigoryk 

> It seems to me like you'll end-up with the same pattern here, but via your own intermediary.

If you're curious why we use this intermediary, it's a part of our MVVM architecture (essentially [the one Google recommends](https://developer.android.com/jetpack/docs/guide#recommended-app-arch)): we represent our model with Observable values that act as single-source-of-truth of declarative state, simplifying our UI updates (like React). Let me know if I should elaborate.

> and UI bits that care about account state can query it and/or register to receive updates. 

These patterns go against our architecture and we've decided against these patterns because:
- Querying introduces multiple sources of truth and update timing errors
- Registering for non-mutually exclusive state updates is 1) non-declarative so generally harder to reason about than a single state object and 2) if multiple components register for the same updates, there can be duplication in assumptions of logic (e.g. if we receive `onAuthenticated`, does each component remember to check if a profile is available?).

tbh, I'm surprised Fenix uses these patterns: iiuc, the MVI-inspired architecture is intended to represent all model state within the state tree, removing the classes of errors I mention above.

> My general recommendation would be to avoid modelling this after internals of the account manager, but rather build on top of data abstractions that it provides

That's my intention, yes – thank you for suggesting it. However, the internal FxA account state seems to be about as minimal as we can get: each of the states needs a separate representation in the UI.

> Generally speaking, any module you add on top of some other module should ideally introduce a higher level of abstraction. If it doesn't, there's a good chance that module isn't needed and just adds complexity.

Agreed. Our architecture does add complexity (through boilerplate and explicitly modeling dependency state) but it introduces simplicity by transforming it into declarative state and allowing us to look at model and UI independently.

btw, I appreciate your suggestions on best practices – keep them coming! 😄